8357445: G1: Time-Based Heap Uncommit During Idle Periods

mo-beck opened the pull request:
**Implements:** https://bugs.openjdk.org/browse/JDK-8357445

Implement time-based heap uncommit for G1 during idle periods.

Key changes:
- Added G1HeapEvaluationTask for periodic heap evaluation
- Switch from G1ServiceTask to PeriodicTask for improved scheduling
- Implemented time-based heap sizing policy with configurable uncommit delay
- Added region activity tracking with last access timestamps
- Integrated VM_G1ShrinkHeap operation for safe heap shrinking
- Added new G1 flags: G1UseTimeBasedHeapSizing, G1TimeBasedEvaluationIntervalMillis, G1UncommitDelayMillis, G1MinRegionsToUncommit
- Added 'sizing' log tag for heap sizing operations

Comprehensive Test Coverage:
- Enhanced TestG1RegionUncommit: minimum heap boundaries, concurrent allocation/uncommit scenarios
- Enhanced TestTimeBasedHeapSizing: humongous object handling, rapid allocation cycles, edge cases
- Enhanced TestTimeBasedRegionTracking: concurrent region access, lifecycle transition validation
- Enhanced TestTimeBasedHeapConfig: parameter boundary values, small heap configurations

This ensures time-based heap uncommit works correctly while maintaining all safety guarantees and test expectations.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Error
&nbsp;⚠️ Pull request body is missing required line: `- [x] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai).`

### Issue
 * [JDK-8357445](https://bugs.openjdk.org/browse/JDK-8357445): G1: Time-Based Heap Uncommit During Idle Periods (**Enhancement** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/26240/head:pull/26240` \
`$ git checkout pull/26240`

Update a local copy of the PR: \
`$ git checkout pull/26240` \
`$ git pull https://git.openjdk.org/jdk.git pull/26240/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 26240`

View PR using the GUI difftool: \
`$ git pr show -t 26240`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/26240.diff">https://git.openjdk.org/jdk/pull/26240.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/26240#issuecomment-3056778470)
</details>

mo-beck committed:
JDK-8357445: G1: Time-Based Heap Uncommit During Idle Periods

Implement time-based heap uncommit for G1 during idle periods.

Key changes:
- Added G1HeapEvaluationTask for periodic heap evaluation
- Switch from G1ServiceTask to PeriodicTask for virtual thread support
- Implemented time-based heap sizing policy with configurable uncommit delay
- Added region activity tracking with last access timestamps
- Integrated VM_G1ShrinkHeap operation for safe heap shrinking
- Added new G1 flags: G1UseTimeBasedHeapSizing, G1TimeBasedEvaluationIntervalMillis, G1UncommitDelayMillis, G1MinRegionsToUncommit
- Added 'sizing' log tag for heap sizing operations

Comprehensive Test Coverage:
- Enhanced TestG1RegionUncommit: minimum heap boundaries, concurrent allocation/uncommit scenarios
- Enhanced TestTimeBasedHeapSizing: humongous object handling, rapid allocation cycles, edge cases
- Enhanced TestTimeBasedRegionTracking: concurrent region access, lifecycle transition validation
- Enhanced TestTimeBasedHeapConfig: parameter boundary values, small heap configurations

This ensures time-based heap uncommit works correctly while maintaining all safety guarantees and test expectations.

mo-beck commented:
/label add hotspot-gc

mo-beck committed:
Normalize encoding and line endings for G1 GC time-based tests

walulyai commented on src/hotspot/share/gc/g1/g1CollectedHeap.cpp:
```diff
@@ -2914,6 +2959,8 @@ void G1CollectedHeap::retire_mutator_alloc_region(G1HeapRegion* alloc_region,
   assert_heap_locked_or_at_safepoint(true /* should_be_vm_thread */);
   assert(alloc_region->is_eden(), "all mutator alloc regions should be eden");
 
+  alloc_region->record_activity();  // Record the activity of the alloc region
```
Since we only uncommit free regions, we only need to `record_activity` when the region is freed which is done in `G1HeapRegion::hr_clear` before the region is added to the free-list

mo-beck replied:
Thanks for pointing this out—you're right, and to clarify, the logic explicitly ensures that we only ever uncommit free regions. When I designed this, I tracked both allocation activities and retirements to record region idle times as accurately and minimally as possible.

For precise time-based tracking, I believe we need to update the timestamp at both `hr_clear()` and retirement. Here’s the timing consideration:

**The timing issue:**
1. Region cleared → `hr_clear()` → `record_activity()` → sets timestamp = T1
2. Region is allocated and used for extended period
3. Region is retired → if we only update the timestamp at clear, not at retirement, there’s no new timestamp
4. Time-based evaluation → sees timestamp T1 (from clear, not retire)
5. Result: Region appears "idle" since T1, even though it was just recently retired

So, we need `record_activity()` at both lifecycle points:

- **`hr_clear()`**: When a region becomes available (initialization)
- **`retire_*()`**: When a region finishes active use (retirement)

This way we always track the **most recent activity** for each region, whether that was clearing or retirement, giving us accurate idle time calculations for time-based uncommit decisions.

Since the logic guarantees we only uncommit regions that are truly free and have been idle for the configured delay, having an accurate “last active” timestamp is critical for the feature’s effectiveness.

walulyai replied:
IIUC `retire` a region means that the region has been retired from allocation, however, the particular region is in active use just that we can no longer allocate out of this region. Eventually, this region will be evacuated and `hr_clear()` called. So `record_activity()` during retirement is redundant.

mo-beck replied:
You're right about removing the redundant calls. Plus, my logic already ensures only empty regions are considered for uncommitting, so retired region timestamps wouldn't affect decisions anyway

mo-beck commented on src/hotspot/share/gc/g1/g1HeapEvaluationTask.cpp:
```diff
@@ -0,0 +1,81 @@
+
+void G1HeapEvaluationTask::task() {
+  // This runs on WatcherThread during idle periods - perfect for time-based evaluation!
+  log_debug(gc, sizing)("Starting heap evaluation");
```
> For existing Heap Resizing tasks, we use the tags `(gc, ergo, heap)`, I would suggest we maintain that.

I'd like to discuss the logging tag choice. While I understand the consistency argument for using (gc,ergo,heap), I believe (gc,sizing) is more appropriate here because:

1. This is a distinct sizing mechanism, not traditional ergonomic heap management
2. Users troubleshooting time-based sizing specifically will benefit from a dedicated tag
3. It aligns with the feature name and purpose
4. It provides better separation of concerns for logging

Would you be open to keeping (gc,sizing) for semantic clarity?

walulyai replied:
My argument is related to consistency, if there are reasons to have this logging distinct from the `(gc, ergo, heap)`, that is fine by me.

mo-beck replied:
Thank you for being open to the distinct logging tags. I'll keep (gc,sizing) to maintain the semantic clarity for time-based heap sizing troubleshooting. This will help users and support teams isolate time-based sizing behavior from traditional ergonomic heap management.

mo-beck commented on src/hotspot/share/gc/g1/g1Allocator.hpp:
```diff
@@ -32,6 +32,7 @@
 
 class G1EvacInfo;
 class G1NUMA;
+class G1HeapSizingPolicy;
```
removed

mo-beck commented on src/hotspot/share/gc/g1/g1Allocator.inline.hpp:
```diff
@@ -54,11 +54,11 @@ inline HeapWord* G1Allocator::attempt_allocation(uint node_index,
-  if (result != nullptr) {
-    return result;
+  if (result == nullptr) {
+    result = mutator_alloc_region(node_index)->attempt_allocation(min_word_size, desired_word_size, actual_word_size);
   }
 
-  return mutator_alloc_region(node_index)->attempt_allocation(min_word_size, desired_word_size, actual_word_size);
+  return result;
```
sounds good, will revert.

mo-beck commented on src/hotspot/share/gc/g1/g1CollectedHeap.cpp:
```diff
@@ -1070,8 +1073,14 @@ void G1CollectedHeap::shrink_helper(size_t shrink_bytes) {
-                            shrink_bytes, shrunk_bytes, num_regions_removed);
+                           shrink_bytes, shrunk_bytes, num_regions_removed);
   if (num_regions_removed > 0) {
+    log_info(gc, heap)("Heap shrink completed: uncommitted %u regions (%zuMB), heap size now %zuMB",
```
changed to "flagged"

mo-beck commented on src/hotspot/share/gc/g1/g1CollectedHeap.cpp:
```diff
@@ -1472,6 +1502,15 @@ jint G1CollectedHeap::initialize() {
+    // PeriodicTask will be enrolled after G1 is fully initialized in post_initialize()
+    log_debug(gc, init)("G1 Time-Based Heap Evaluation task created (PeriodicTask)");
+  } else {
+    _heap_evaluation_task = nullptr;
```
replaced with assert

mo-beck commented on src/hotspot/share/gc/g1/g1CollectedHeap.cpp:
```diff
@@ -1531,6 +1570,12 @@ void G1CollectedHeap::safepoint_synchronize_end() {
   ref_processing_init();
+
+  // Enroll the heap evaluation task after G1 is fully initialized
+  if (G1UseTimeBasedHeapSizing && _heap_evaluation_task != nullptr) {
```
fixed - when G1UseTimeBasedHeapSizing is true, _heap_evaluation_task is created in the constructor, so it should never be nullptr in post_initialize()

mo-beck commented on src/hotspot/share/gc/g1/g1CollectedHeap.hpp:
```diff
@@ -918,6 +927,8 @@ class G1CollectedHeap : public CollectedHeap {
   // The current policy object for the collector.
   G1Policy* policy() const { return _policy; }
+  // The heap sizing policy.
+  G1HeapSizingPolicy* heap_sizing_policy() const { return _heap_sizing_policy; }
```
removed

mo-beck commented on src/hotspot/share/gc/g1/g1HeapSizingPolicy.cpp:
```diff
@@ -23,14 +24,28 @@
+
+void G1HeapSizingPolicy::initialize() {
+  // Flag values are available at this point
+  _uncommit_delay_ms = (jlong)G1UncommitDelayMillis;
```
Good point! You're absolutely right - since G1HeapSizingPolicy uses the singleton pattern through `create()`, we can safely move the initialization into the constructor and eliminate the separate `initialize()` method. Made the change:

- Moved `_uncommit_delay_ms` initialization into the constructor
- Removed the `initialize()` method
- Removed the call to `initialize()` from G1CollectedHeap constructor

Much cleaner approach, thanks for the suggestion!

mo-beck commented on src/hotspot/share/runtime/vmOperations.cpp:
```diff
@@ -638,6 +641,12 @@ void VM_PrintCompileQueue::doit() {
+  _g1h->shrink(_bytes);
+}
+#endif // INCLUDE_G1GC
+
```
Good suggestion! Moved `VM_G1ShrinkHeap::doit()` implementation from `vmOperations.cpp` to `g1VMOperations.cpp` where it belongs with the other G1-specific VM operations. Much better organization.

mo-beck commented on src/hotspot/share/runtime/vmOperations.hpp:
```diff
@@ -291,4 +293,19 @@ class VM_PrintClassHierarchy: public VM_Operation {
 };
 #endif // INCLUDE_SERVICES
 
+#if INCLUDE_G1GC
+class VM_G1ShrinkHeap : public VM_Operation {
+ private:
+  G1CollectedHeap* _g1h;
+  size_t _bytes;
+ public:
+  VM_G1ShrinkHeap(G1CollectedHeap* g1h, size_t bytes)
+    : _g1h(g1h), _bytes(bytes) {}
+  VMOp_Type type() const override { return VMOp_G1ShrinkHeap; }
+  const char* name() const override { return "G1ShrinkHeap"; }
+  bool is_gc_operation() const override { return true; }
+  void doit() override;
+};
```
done.

mo-beck committed:
8357445: Address code review feedback

Extra line removed
Include cycle comment removed
completed changed to flagged
nullptr assignment changed to assert
Redundant nullptr check removed
VM_G1ShrinkHeap moved
Forward declaration removed
Constructor initialization for static _uncommit_delay_ms

mo-beck committed:
8357445: Remove redundant record_activity calls and leftover initialize call

- Remove record_activity() from retirement methods as hr_clear() is always last
- Remove leftover initialize() call since initialization moved to constructor
- Remove unused G1 includes from vmOperations after moving VM_G1ShrinkHeap

tschatzl commented on src/hotspot/share/gc/g1/g1CollectedHeap.cpp:
```diff
@@ -1116,6 +1124,24 @@ void G1CollectedHeap::shrink(size_t shrink_bytes) {
+  // GC service thread) just do the work directly.
+  if (SafepointSynchronize::is_at_safepoint()) {
+    shrink(shrink_bytes);
+    return true;                     // we *did* something
```
(Full) Sentence length comments should follow sentence style: Initial capitalization and full stop at the end. Not sure if the emphasis should be kept, seems to be overused (and I do not think we use it a lot in G1 code).
(Please no markdown btw :) )

tschatzl commented on src/hotspot/share/gc/g1/g1HeapEvaluationTask.cpp:
```diff
@@ -0,0 +1,81 @@
+  if (resize_amount > 0) {
+    // Time-based evaluation only handles uncommit/shrinking, never expansion
+    if (should_expand) {
+      log_warning(gc, sizing)("Time-based evaluation unexpected expansion request ignored (resize_amount=%zuB)", resize_amount);
```
I think "Time-based evaluation" as a kind of extra tag seems to be too unspecific: Time based evaluation of what?

I.e. the heap sizing based on GCTimeRatio is also "time-based". Also the messages partially seem to be inconsistent at first glance: Somebody this "time-based" prefix is there, sometimes not. Also sometimes it reads "Time-based evaluation", sometimes "Time-based heap sizing", sometimes having ":" in them, sometimes not.

If we add an extra tag, that "Time-based*" prefix might be superfluous, particularly if we keep the `sizing`for this kind of stuff (or even make it more specific).

tschatzl commented on src/hotspot/share/gc/g1/g1HeapSizingPolicy.hpp:
```diff
@@ -66,6 +71,8 @@ class G1CollectedHeap;
 // For full collections, we base resize decisions only on Min/MaxHeapFreeRatio.
 //
 class G1HeapSizingPolicy: public CHeapObj<mtGC> {
+  static jlong _uncommit_delay_ms;  // Delay before uncommitting inactive regions
```
It's generally better to use `Ticks`/`Timespan` etc for times as they a) have higher precision and b) one does not need to carry the unit in them. Just convert when you need. They are also distinct types so that wrong assignment can not happen (like a value containing seconds to something that ought to contain milliseconds).
One can only round down etc. when needed.

Also, `jlong` should be avoided, as it is originally a "J"ava type, i.e. for Java interfacing (yeah, it has crept in over time for various reasons).

mo-beck replied:
Great points. Will update to:

```
Ticks current_ticks = Ticks::now();
Tickspan elapsed = current_ticks - _last_access_ticks;
return elapsed > _uncommit_delay;
```

tschatzl commented on src/hotspot/share/gc/g1/g1HeapSizingPolicy.cpp:
```diff
@@ -436,3 +448,150 @@ size_t G1HeapSizingPolicy::full_collection_resize_amount(bool& expand, size_t al
+    // Calculate maximum bytes we can uncommit while respecting min heap size
+    size_t max_shrink_bytes = current_heap > min_heap ? current_heap - min_heap : 0;
+
+    log_trace(gc, sizing)("Time-based evaluation details: current_heap=%zuB min_heap=%zuB "
```
I do not think `trace`log messages need the "details" in `Time-based evaluation details" - this is already indicated by the log level.

Also, in contrast to other "Time-based evaluation" messages this has a ":" at the end, which is inconsistent.

mo-beck replied:
ok changed to:
log_trace(gc, sizing)("Uncommit evaluation: current_heap=%zuB min_heap=%zuB ...")

tschatzl reviewed (commented):
Some initial drive-by comments...

tschatzl commented on src/hotspot/share/gc/g1/g1_globals.hpp:
```diff
@@ -370,6 +370,25 @@
           "scan cost related prediction samples. A sample must involve "    \
           "the same or more than this number of code roots to be used.")    \
                                                                             \
+  product(bool, G1UseTimeBasedHeapSizing, false, EXPERIMENTAL,              \
+          "Enable time-based heap sizing to uncommit memory from inactive " \
+          "regions independent of GC cycles")                               \
+                                                                            \
+  product(uintx, G1TimeBasedEvaluationIntervalMillis, 60000, MANAGEABLE,    \
+          "Interval in milliseconds between periodic heap-size evaluations "\
+          "when G1UseTimeBasedHeapSizing is enabled")                       \
+          range(1000, LP64_ONLY(max_jlong) NOT_LP64(max_uintx / 2))         \
+                                                                            \
+  product(uintx, G1UncommitDelayMillis, 300000, MANAGEABLE,                 \
+          "A region is considered inactive if it has not been accessed "    \
+          "within this many milliseconds")                                  \
+          range(1000, LP64_ONLY(max_jlong) NOT_LP64(max_uintx / 2))         \
+                                                                            \
+  product(size_t, G1MinRegionsToUncommit, 10, EXPERIMENTAL,                 \
+          "Minimum number of inactive regions required before G1 will "     \
+          "attempt to uncommit memory")                                     \
+          range(1, max_uintx)                                               \
+                                                                            \
```
One reason why I do not see the need for these flags to be manageable is that this heap shrinking is a supportive function if the application is "idle". Main heap sizing should be done by the existing heap sizing/AHS at GC events.

From the CR text:
>  Doing heap sizing based on garbage collections has a big disadvantage: if there are no garbage collections (due to no application activity) there is a risk that a large amount of heap is kept committed unnecessarily *for a long time*. 
(emphasis mine)

If we are talking about "a long time", so there does not seem to be a need for changing it during runtime (or change it at all). It should not matter if that "long time" is "long time +- small epsilon", and so, allowing dynamic change of "a long time" to another "long time" seems undesirable without a very good use case.

Another question is whether you had thoughts about the interaction with (JDK-8213198)[https://bugs.openjdk.org/browse/JDK-8213198], as this change seems to be a subset of the other. (That's just curiosity from me, I think this feature is useful as is, and if the other ever materializes, we can always reconsider).

Otoh decreasing the heap by this mechanism will eventually trigger a marking.

walulyai commented on src/hotspot/share/gc/g1/g1HeapEvaluationTask.hpp:
```diff
@@ -0,0 +1,47 @@
+
+// Time-based heap evaluation task that runs during idle periods.
+// Uses PeriodicTask rather than G1ServiceTask due to build compatibility issues
+// in JDK 25+. PeriodicTask's 10ms granularity is adequate for heap evaluation
```
Can you please elaborate on these "compatibility issues"

mo-beck replied:
changed

walulyai commented:
Hi,

There appears to be a disconnect between the `get_uncommit_candidates` logic and the actual heap shrinking performed by `G1CollectedHeap::shrink`. While `G1HeapSizingPolicy::evaluate_heap_resize` determines the number of bytes to shrink (via shrink_bytes) and passes this to the heap shrink logic, the regions identified as uncommit candidates are not explicitly communicated or prioritized during the shrink operation.

As a result, the heap may be shrunk without necessarily uncommitting the specific regions previously marked as uncommit candidates. This can lead to a scenario where those regions remain committed even after the shrink, potentially triggering repeated shrink attempts in subsequent calls to `G1HeapSizingPolicy::evaluate_heap_resize`.

Is this understanding correct?

mo-beck committed:
JDK-8357445: Implement @tschatzl code review feedback for time-based heap sizing

- Convert from jlong milliseconds to proper time types (Ticks/Tickspan)
- Reclassify G1UseTimeBasedHeapSizing from EXPERIMENTAL+false to DIAGNOSTIC+true
- Reclassify G1MinRegionsToUncommit from EXPERIMENTAL to DIAGNOSTIC
- Keep timing flags (G1UncommitDelayMillis, G1HeapEvaluationIntervalMillis) as MANAGEABLE
- Update all test files to use UnlockDiagnosticVMOptions instead of UnlockExperimentalVMOptions
- Remove explicit G1UseTimeBasedHeapSizing=true from tests (now enabled by default)
- Improve logging terminology from 'time-based' to 'uncommit evaluation'
- Fix comment style and remove unused heap_sizing_policy() accessor
- Comprehensive type safety improvements throughout the codebase

This addresses all feedback regarding type safety, flag design,
and production readiness of the time-based heap sizing feature.

mo-beck committed:
JDK-8357445: Implement @tschatzl code review feedback for time-based heap sizing

- Convert from jlong milliseconds to proper time types (Ticks/Tickspan)
- Reclassify G1UseTimeBasedHeapSizing from EXPERIMENTAL+false to DIAGNOSTIC+true
- Reclassify G1MinRegionsToUncommit from EXPERIMENTAL to DIAGNOSTIC
- Keep timing flags (G1UncommitDelayMillis, G1HeapEvaluationIntervalMillis) as MANAGEABLE
- Update all test files to use UnlockDiagnosticVMOptions instead of UnlockExperimentalVMOptions
- Remove explicit G1UseTimeBasedHeapSizing=true from tests (now enabled by default)
- Improve logging terminology from 'time-based' to 'uncommit evaluation'
- Fix comment style and remove unused heap_sizing_policy() accessor
- Comprehensive type safety improvements throughout the codebase

This addresses all feedback regarding type safety, flag design,
and production readiness of the time-based heap sizing feature.

mo-beck committed:
Resolve merge conflict - keep proper @tschatzl time type implementation

- Keep Tickspan::from_milliseconds(G1UncommitDelayMillis) initialization
- Keep (jlong)elapsed.milliseconds() cast for macOS build compatibility
- Keep elapsed > _uncommit_delay direct Tickspan comparison
- Maintains cross-platform compatibility

mo-beck committed:
Remove unused static _uncommit_delay member and accessor

tschatzl reviewed (changes requested):
I did not look through the tests yet, there are quite a few areas that need fixes first.

tschatzl commented:
Also, garbage collections and the `VM_G1ShrinkHeap` operation need to reset the timestamp for all (free) regions too. Otherwise the mechanism will try to uncommit the wrong amount of regions, similar to the issue I described in `VM_G1ShrinkHeap`.

E.g. Regions A, B, C are free for a long time, GC happens that keeps A, B, C for various reasons, now idle-evaluation happens, and A, B, C are determined idle in addition to the other regions if their timestamp is not reset. So the idle uncommit will free too many regions.

mo-beck commented on src/hotspot/share/gc/g1/g1VMOperations.cpp:
```diff
@@ -170,3 +170,7 @@ void VM_G1PauseCleanup::work() {
 }
+
+void VM_G1ShrinkHeap::doit() {
+  _g1h->shrink(_bytes);
```
I have added execution-time handling that uses a pre-evaluated shrink amount but performs fresh region selection at VM operation execution time:

`_g1h->shrink_with_time_based_selection(_bytes);`

At execution time, we do the calculation/fresh eval by clearing the cache and applying the time-based criteria:

```
void G1CollectedHeap::shrink_with_time_based_selection(size_t shrink_bytes) {
  // Fresh region scanning and validation at VM operation execution time
  _hrm.remove_all_free_regions();
  shrink_helper_with_time_based_selection(aligned_shrink_bytes);
  rebuild_region_sets(true);
}
```
The actual region selection:

`num_regions_removed = _hrm.shrink_by(num_regions_to_remove, true /* use_time_based_selection */);`

This should address the stale region selection concern (A,B,C → D,E,F) by ensuring region eligibility is determined fresh at VM operation execution time rather than using potentially outdated information from the scheduling phase.

Commit:
Address OpenJDK PR feedback on time-based heap uncommit and region selection

Commit:
Merge master

Commit:
8357445: Fix trailing whitespace issues in G1 time-based heap sizing implementation

- Remove trailing whitespace from all modified G1 files
- Resolves OpenJDK jcheck whitespace validation errors
- No functional changes, whitespace cleanup only

Commit:
8357445: Address feedback for G1 time-based heap sizing

- Fix indentation in log_debug statement in shrink_helper (suggested by @tschatzl)
- Change terminology from 'inactive' to 'idle' throughout time-based heap sizing
- Update flag descriptions in g1_globals.hpp to use 'idle' terminology
- Fix remaining trailing whitespace in test files

Addresses all outstanding review comments from @tschatzl

Commit:
Fix compilation errors after master merge

- Fix syntax error in g1CollectedHeap.cpp (extra closing brace)
- Update API call from short_term_pause_time_ratio() to short_term_gc_time_ratio()

These fixes resolve compilation issues that occurred after merging with
upstream master due to API changes in the OpenJDK codebase.

mo-beck commented:
/touch

tschatzl commented on src/hotspot/share/gc/g1/g1HeapEvaluationTask.cpp:
```diff
@@ -0,0 +1,72 @@
+void G1HeapEvaluationTask::execute() {
+  log_debug(gc, sizing)("Starting uncommit evaluation");
+
+  ResourceMark rm; // Ensure temporary resources are released
```
Which ones are those?

tschatzl commented on src/hotspot/share/gc/g1/g1CollectedHeap.cpp:
```diff
@@ -1090,23 +1095,99 @@ bool G1CollectedHeap::expand_single_region(uint node_index) {
+
+  // We should only reach here from the service thread during idle time
+  // but ensure any GC alloc regions are abandoned
+  _allocator->abandon_gc_alloc_regions();
```
Maybe check that they are abandoned (i.e. there are no gc alloc regions) in an assert instead of doing work.

tschatzl commented on src/hotspot/share/gc/g1/g1CollectedHeap.cpp:
```diff
@@ -1090,23 +1095,99 @@ bool G1CollectedHeap::expand_single_region(uint node_index) {
+                       num_regions_removed, shrunk_bytes / M, capacity() / M);
+    log_debug(gc, heap)("Heap shrink details: requested=%zuB actual=%zuB "
+                        "regions_removed=%u heap_capacity=%zuB",
+                        shrink_bytes, shrunk_bytes, num_regions_removed, capacity());
```
It would be nice to not repeat the same information at different levels; one can check the current level with `LogTarget::is_enabled()`.

tschatzl commented on src/hotspot/share/gc/g1/g1CollectedHeap.hpp:
```diff
@@ -992,6 +1007,9 @@ class G1CollectedHeap : public CollectedHeap {
 
+  // Deactivate a specific region by index.
+  void deactivate_region_at(uint region_index) { _hrm.shrink_at(region_index, 1); }
+
```
Unused. Remove.

tschatzl commented on src/hotspot/share/gc/g1/g1HeapRegion.cpp:
```diff
@@ -249,6 +250,7 @@ G1HeapRegion::G1HeapRegion(uint hrm_index,
   _surv_rate_group(nullptr),
   _age_index(G1SurvRateGroup::InvalidAgeIndex),
   _node_index(G1NUMA::UnknownNodeIndex),
+  _last_access_timestamp(Ticks::now()), // Initialize timestamp with current time
```
This call is still unnecessary. The constructor will call `initialize` which calls `hr_clear()` which does the initialization. Just set it to zero (or some "invalid value" here like other members show).

tschatzl commented on src/hotspot/share/gc/g1/g1HeapRegion.cpp:
```diff
@@ -156,6 +157,7 @@ void G1HeapRegion::clear_both_card_tables() {
 void G1HeapRegion::set_free() {
   if (!is_free()) {
     report_region_type_change(G1HeapRegionTraceType::Free);
+    record_activity(); // Record timestamp when region becomes free
```
Make this unconditional. At startup, the default value of the region's type is `Free`, so this will not show up if removing the call to `hr_clear()`.

tschatzl commented on src/hotspot/share/gc/g1/g1HeapRegionManager.hpp:
```diff
@@ -283,6 +287,10 @@ class G1HeapRegionManager: public CHeapObj<mtGC> {
+  // Reset access timestamps on all free regions to prevent stale timing data
+  // Used after GC operations or shrink operations that may affect region state
+
+
```
Leftover comment?

tschatzl commented on src/hotspot/share/gc/g1/g1HeapRegionManager.cpp:
```diff
@@ -596,25 +601,97 @@ uint G1HeapRegionManager::shrink_by(uint num_regions_to_remove) {
+
+    shrink_at(region_index, 1);
+    removed++;
+  }
```
Also the code *must* reset the timestamp for regions not selected for the reasons stated in another comment.

tschatzl commented on src/hotspot/share/gc/g1/g1HeapSizingPolicy.cpp:
```diff
@@ -436,3 +446,262 @@ size_t G1HeapSizingPolicy::full_collection_resize_amount(bool& expand, size_t al
+  // Back off during allocation pressure - only evaluate when truly idle
+  if (_analytics != nullptr) {
+    double gc_time_ratio = _analytics->short_term_gc_time_ratio();
+    if (gc_time_ratio > 0.05) { // 5% GC time still indicates pressure
```
gc time ratio is no indicator for idleness. Also the default is 4% right now (ie. which G1 aims for), this value would be higher than the goal. Also the time ratio is user-settable - some users may accept a higher gc usage than others.

I understand the need for this change to kind of see if there is "pressure" - however this would be implicit by gcs occurring inbetween (i.e. if the `ServiceTask` did not occur/did not find need for uncommitting regions between GCs due to time stamps too high, then there is nothing to do anyway).

tschatzl reviewed (changes requested):
I stopped reviewing after >40 issues (still finding some while writing) because
- some of them were unresolved from last time. Please resolve comments that have been addressed or comment why the change did not address it. Having old unresolved comments are annoying for re-reviews because they make the reviewer unsure if there is some need for further clarification or not.
- new code introduced the same issues again (e.g. use of `is_empty()`) without comment on why it can be used, after giving the reason why it can't be used.
- the main issue that the time-based uncommit still interferes with gc based sizing without restriction still applies.
- another is that the VM operation still does not re-evaluate the decisions, just uncommits a set number of regions. Possibly not even those that were found as candidates earlier afaict.
- the only improvement I remember has been that `ServiceTask` is used now

mo-beck committed:
Address Thomas Schatzl's review comments for JDK-8357445

This commit addresses all issues raised in Thomas's November 2025 review
of the time-based heap uncommit feature (PR #26240).

Major architectural fix:
- VM_G1ShrinkHeap::doit() now re-evaluates uncommit candidates at safepoint
  time by calling find_uncommit_candidates_by_time() during VM operation
  execution. This ensures regions are validated as still free at the point
  of uncommit, fixing the race condition where heap state could change
  between service thread evaluation and VM operation execution.

Code quality improvements:
- Removed redundant is_empty() checks in g1HeapSizingPolicy.cpp and
  g1HeapRegionManager.cpp since is_free() already implies empty
- Replaced inefficient bubble sort with GrowableArray::sort() using
  lambda comparator for O(n log n) performance
- Fixed lambda signature (G1HeapRegion** instead of const*) to match
  GrowableArray::sort() requirements
- Added periods to sentence-style comments for consistency

The implementation already had correct:
- SuspendibleThreadSetJoiner placement in service thread task
- GC coordination with 5% overhead threshold and G1ReservePercent
- Proper assertions and validations

Tested with SPECjbb2015 showing correct time-based region selection
and uncommit behavior across varying load patterns.

mo-beck committed:
Merge master into feature branch

mo-beck committed:
8357445: Address review feedback with O(1) timestamp reset

- Implement O(1) global timestamp reset
- Add _last_gc_timestamp member for baseline tracking
- Call reset_free_region_timestamps() after young and full GC
- Use MAX2(region_time, _last_gc_time) for effective timestamp calculation
- Change request_heap_shrink() return type to void
- Remove unused deactivate_region_at() method
- Simplify comments to follow conventions
- Remove redundant ResourceMark and duplicate logging

mo-beck commented:
Hi @tschatzl 

Following up on my December reply with additional fixes and performance verification.

## Fixes

| Issue | File | Status |
|-------|------|--------|
| O(1) timestamp reset after GC | g1HeapRegionManager | Added `_last_gc_timestamp`; `reset_free_region_timestamps()` called after young/full GC; uses `MAX2(region_time, _last_gc_time)` |
| `_last_access_timestamp(Ticks::now())` in constructor | g1HeapRegion.cpp | Default-initialized (0); set when region transitions to free |
| `record_activity()` call in hr_clear() | g1HeapRegion.cpp | Removed - `set_free()` handles timestamp update |
| `record_activity()` method | g1HeapRegion.hpp | Renamed to `update_last_access_timestamp()` |
| Unused `should_uncommit()` method | g1HeapRegion.hpp | Removed |
| Unused `deactivate_region_at()` method | g1CollectedHeap.hpp | Removed |
| `_heap_evaluation_task` not in initializer list | g1CollectedHeap.cpp | Moved to constructor initializer list |
| Unnecessary assert about `_heap_evaluation_task` | g1CollectedHeap.cpp | Removed |
| Unnecessary init log when registering task | g1CollectedHeap.cpp | Removed (G1ServiceThread handles it) |
| `request_heap_shrink()` return value never used | g1CollectedHeap | Changed to `void` |
| `total_regions` variable unused | g1HeapSizingPolicy.cpp | Removed |
| `current_heap` naming unclear | g1HeapSizingPolicy.cpp | Renamed to `current_capacity` |
| `_analytics` could be nullptr | g1HeapSizingPolicy.cpp | Added `assert(_analytics != nullptr)` in constructor |
| G1ReservePercent calculation adds instead of MAX | g1HeapSizingPolicy.cpp | Uses `MAX2(g1_reserve_regions, young_gen_regions)` |
| ResourceMark unnecessary | g1HeapSizingPolicy.cpp | Removed |
| Duplicate log messages | g1HeapEvaluationTask.cpp | Consolidated info/debug into single log |
| Manual iteration instead of `heap_region_iterate()` | g1HeapRegionManager.cpp | Uses closure with `iterate()` |
| `current_time` should be taken once | g1HeapRegionManager.cpp | Taken before iteration |

## Performance Verification

SPECjbb2015 benchmarking (4 iterations per configuration) across 5 TBS configurations varying `G1UncommitDelayMillis` and `G1TimeBasedEvaluationIntervalMillis`. No performance regression when compared to baseline (default).

## Lower Priority Items

Happy to address if you prefer:
- Task naming: `G1HeapEvaluationTask` → `G1TimeBasedHeapEvaluationTask`
- Log context: Add more detail at info level

Regarding log tags: I'd prefer to keep `gc,sizing` as TBS is a distinct feature - it's not STW or concurrent GC, just memory reclaim. The `sizing` tag helps users isolate TBS behavior from traditional GC logging.

Looking forward to your thoughts.

Thanks,
Monica

mo-beck committed:
Merge branch 'master' into feature/JDK-8357445-time-based-heap-sizing

mo-beck commented on src/hotspot/share/gc/g1/g1HeapEvaluationTask.cpp:
```diff
@@ -0,0 +1,81 @@
+#include "utilities/debug.hpp"
+#include "utilities/globalDefinitions.hpp"
+
+G1HeapEvaluationTask::G1HeapEvaluationTask(G1CollectedHeap* g1h, G1HeapSizingPolicy* heap_sizing_policy) :
```
Keeping this as is for now. If you all feel strongly about it, I can rename it to `G1HeapUncommitEvaluationTask` or `G1TimeBasedHeapShrinkTask`.

mo-beck committed:
8357445: Address Thomas Schatzl review feedback

VM_G1ShrinkHeap inherits VM_GC_Operation with gc_count_before
coordination. Rename inactive to idle, inline timestamp update,
simplify find_uncommit_candidates_by_time by removing redundant
max_candidates parameter, fix reserved regions calculation to sum
young gen + reserve buffer. Clean up comment style, log prefixes,
remove stale include and unnecessary casts.
